src/main/java/com/example/springlabs/controllers/CategoryController.java

```diff
@@ -0,0 +1,94 @@
+package com.example.springlabs.controllers;
+
+import com.example.springlabs.controllers.dtos.CategoryDto;
+import com.example.springlabs.controllers.dtos.mappers.CategoryDtoMapper;
+import com.example.springlabs.services.CategoryService;
+import jakarta.servlet.http.HttpServletRequest;
+import java.net.URLDecoder;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import lombok.AccessLevel;
+import lombok.RequiredArgsConstructor;
+import lombok.experimental.FieldDefaults;
+import org.springframework.http.HttpStatus;
+import org.springframework.util.AntPathMatcher;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.servlet.HandlerMapping;
+
+@RestController
+@RequestMapping("/categories")
+@RequiredArgsConstructor
+@FieldDefaults(level = AccessLevel.PRIVATE, makeFinal = true)
+public class CategoryController {
+
+  CategoryService categoryService;
+
+  CategoryDtoMapper categoryDtoMapper;
+
+  @GetMapping
+  public Collection<CategoryDto> getAll() {
+    return categoryDtoMapper.createDtosSet(categoryService.getAllCategories());
+  }
+
+  @GetMapping("/**")
+  public Collection<CategoryDto> getSubcategories(HttpServletRequest request) {
+    List<String> categoriesNames = getCategoriesNames(request);
+    return categoryDtoMapper.createDtosSet(categoryService.getSubcategories(categoriesNames));
+  }
+
+  @PostMapping
+  @ResponseStatus(HttpStatus.CREATED)
+  public CategoryDto addCategory(@RequestBody CategoryDto categoryDto) {
```

Comment:
Mb add smth like CategoryCreateDto without products and subCategories fields. I think new category immediatly after its creation does not have subCategories and products